Raised IOException on deleteBlob

gfyoung opened the pull request:
Title is self-explanatory.  Closes #18530.

gfyoung commented:
Suggestions on how to write tests for these changes?

jasontedor commented on core/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java:
```diff
@@ -85,7 +86,13 @@ public FsBlobContainer(FsBlobStore blobStore, BlobPath blobPath, Path path) {
+                throw new IOException("File " + blobPath.toString() + " does not exist");
+            }
+        } catch (DirectoryNotEmptyException | SecurityException e) {
+            throw new IOException(e);
```
I don't think that a security exception should be re-thrown as an `IOException`.

jasontedor replied:
A `DirectoryNotEmptyException` is an `IOException`, why is it being wrapped in an `IOException`?

gfyoung replied:
See my question in my comment <a href="https://github.com/elastic/elasticsearch/issues/18530#issuecomment-225205087">here</a>.

jasontedor replied:
I think it's mistake to wrap a `SecurityException` in an `IOException`, I think that the Javadocs need to changed. And again, a `DirectoryNotEmptyException` is already an `IOException` it doesn't need to wrapped and it allow it to bubble up does not violate what should be the contract for this method.

gfyoung replied:
Regarding the DirectoryNotFoundException, my bad.  I misread your comment there.  With regards to the SecurityException, I do see what you mean, though probably input from @abeyad

abeyad replied:
We should not catch the `SecurityException` at all.  Let it propagate.  We should not have even gotten to this point if the security manager did not give us access here, but in any case, its not an exception we should handle at this level.  It should just be propagated.

jasontedor replied:
> We should not catch the `SecurityException` at all. Let it propagate.

Precisely.

jasontedor commented on plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsBlobContainer.java:
```diff
@@ -75,8 +75,8 @@ public Boolean run(FileContext fileContext) throws IOException {
-        } catch (FileNotFoundException ok) {
-            // behaves like Files.deleteIfExists
+        } catch (FileNotFoundException e) {
+            throw new IOException(e);
```
A `FileNotFoundException` is an `IOException`, why wrap it?

gfyoung replied:
See my question in my comment <a href="https://github.com/elastic/elasticsearch/issues/18530#issuecomment-225205087">here</a>.

jasontedor replied:
It's already an `IOException`, it doesn't violate what should be the contract for this method.

gfyoung replied:
Ah, yes that is true.  It was being caught initially (before this PR), so I'll remove the try-catch then.

jasontedor commented:
> Title is self-explanatory.

It explains the what, but not the why and the why is really important.

gfyoung commented:
@jasontedor : I understand that these exceptions are not `IOException`, but then is the issue I'm closing a real issue then?  Also, if you look at the two implementations that already throw `IOException` (see <a href="https://github.com/elastic/elasticsearch/issues/18530#issuecomment-225163982">here</a>), those caught exceptions are not `IOException` either.

gfyoung commented:
@jasontedor : I think the issue that I am closing explains the importance.  Do I nee to repeat it in the PR?

abeyad commented:
@gfyoung Regarding writing tests, take a look at `FsBlobStoreContainerTests`.  It extends `ESBlobStoreContainerTestCase` which contains useful general `BlobContainer` tests that can be used for each of the different `BlobContainer` implementations.

jasontedor commented:
> I understand that these exceptions are not `IOException`, but then is the issue I'm closing a real issue then?

Yes, the leniency is the real issue.

jasontedor commented:
> I think the issue that I am closing explains the importance. Do I need to repeat it in the PR?

Yes, it needs to be in the commit message which ends up as the default body for the PR.

gfyoung commented:
@jasontedor : Regarding the leniency, sure thing.  I suppose a more careful examination of the possible exceptions thrown needs to be done here.  Regarding the commit body, fair enough.  That can be added.

gfyoung commented:
Added a `deleteBlob` test for the `IOException`.  Also added `BlobStoreContainer` tests `AzureBlobContainer`, `HdfsBlobContainer`, and `S3BlobContainer` but I had issues writing their `newBlobStore` methods, especially as simply as the `GoogleCloudStorageBlobStoreContainerTests` for example.  Feedback on how to properly write them would be great!

abeyad commented on plugins/repository-hdfs/src/test/java/org/elasticsearch/repositories/hdfs/HdfsBlobStoreContainerTests.java:
```diff
@@ -0,0 +1,33 @@
+
+    @Override
+    protected BlobStore newBlobStore() throws IOException {
+        return new HdfsBlobStore(FileContext.getFileContext(), "", 100);
```
I don't think this file context is initialized properly.  Look at `HdfsRepository` for how it is initialized, with the appropriate security checks.  You will need to create a temp dir for this, see the `createTempDir()` method that `ESBlobStoreContainerTestCase` inherits.  

Did you run this test?  What did it say, if anything?

gfyoung replied:
It fails at the moment because of permissions.  However I was pretty sure I didn't properly initialize everything either.

abeyad replied:
Yeah I figured... my feeling is these tests weren't written before because they require mocking of the S3/Azure/HDFS services (HDFS may actually work if you give it a `createTempDir()` and initialize it the same way `HdfsRepository` does).  I will be interested to here @imotov and @tlrx thoughts.  The Google blob store repository has a mock storage client that mocks the behavior of what a real Google storage client would do, without actually connecting to an external service.  Something similar would need to be done for S3 and Azure.

gfyoung replied:
Yeah, I was looking for some sort of mock objects as for the Google object, but I could not find any such thing.  Should probably see if I can Mockito somehow.

abeyad replied:
Right, for example, in the case of S3, you would have to create a mock
implementation for the AmazonS3 interface.  Its a pretty lengthy interface,
but you would only need to provide mock implementations for the methods
that S3BlobStore uses, which are just a few.  The rest of the methods could
just return null or some dummy values.

On Sun, Jun 12, 2016 at 12:45 PM, gfyoung <notifications@github.com> wrote:

> In
> plugins/repository-hdfs/src/test/java/org/elasticsearch/repositories/hdfs/HdfsBlobStoreContainerTests.java
> <https://github.com/elastic/elasticsearch/pull/18815#discussion_r66723801>
> :
>
> > + * specific language governing permissions and limitations
> > + * under the License.
> > + */
> > +package org.elasticsearch.repositories.hdfs;
> > +
> > +import org.apache.hadoop.fs.FileContext;
> > +import org.elasticsearch.common.blobstore.BlobStore;
> > +import org.elasticsearch.repositories.ESBlobStoreContainerTestCase;
> > +
> > +import java.io.IOException;
> > +
> > +public class HdfsBlobStoreContainerTests extends ESBlobStoreContainerTestCase {
> > +
> > +    @Override
> > +    protected BlobStore newBlobStore() throws IOException {
> > +        return new HdfsBlobStore(FileContext.getFileContext(), "", 100);
>
> Yeah, I was looking for some sort of mock objects as for the Google
> object, but I could not find any such thing. Should probably see if I can
> Mockito somehow.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/elasticsearch/pull/18815/files/5ecfabc0dffffcbec1cbf7c5c7db0e3ab05a8c07#r66723801>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABjkQVZ4QGKT05XXuqbFiz4DONPvQmbGks5qLDeUgaJpZM4Iy-tw>
> .
>

abeyad commented on plugins/repository-s3/src/test/java/org/elasticsearch/cloud/aws/blobstore/S3BlobStoreContainerTests.java:
```diff
@@ -0,0 +1,46 @@
+
+        return new S3BlobStore(settings, client, bucket, region, false,
+            new ByteSizeValue(100), maxRetries, "public-read-write", "standard");
+    }
```
Again, you can look at how the `S3Repository` initializes the `S3BlobStore`, but we won't be able to run tests like these ... it depends on external S3 access.  I'm not aware of an S3 repository that we have set up explicitly to run such tests.  In this case, I think you would have to write an S3 service mock for tests, which will require more effort.

@imotov may have some ideas here

imotov replied:
@abeyad I think the idea behind ESBlobStoreContainerTestCase was to test the actual (not mocked, unless we can guarantee that mock works exactly like the real store) behavior of the store on the real storage. We cannot run it always, but it should be possible to run it from time to time in a special CI instance to make sure, for example, the behavior of the actual store or its client didn't change on us. To prevent these test from running under normal condition we annotate them with `@ESIntegTestCase.ThirdParty` and provide special configuration to run them

gfyoung replied:
@imotov: Is that the case for the Google repository?  And if it is to test the real behaviour, how to configure this test then if we don't have an S3 repository for testing purposes?  Also, we have one for HDFS or Azure?

imotov replied:
As far as I know, we do have S3 repository for testing purposed. @dadoonet can provide more details. We should get HDFS and Azure as well, if we don't have them already.

gfyoung replied:
@imotov : That would be great to have, since that would facilitate the test setup for those repositories.  However, I would still keep mocking on the table as @abeyad proposed, since that's what the Google cloud tests do with `MockHttpTransport`.

abeyad replied:
Looking further, we actually have an AmazonS3 test class already, look
at: org.elasticsearch.cloud.aws.TestAmazonS3.  Its used in
TestAwsS3Service, which as far as I can tell, is unused.  TestAmazonS3
looks to provide what you would need for creating a real test against S3,
as @imotov mentioned.  AzureBlobStoreTests already has this implemented,
which could be used for a container class test.  Looking at the Google mock
a bit more closely, it seems to be testing against the google APIs
themselves.

If we are testing against real repositories from time to time, it probably
takes away the need to mock the blob store implementations themselves.  In
that case, we should just implement a test against the real back-end
storage mechanism, and run it before finalizing this PR.  It seems like the
infrastructure for this is already in place, though I'm not sure where the
TestAwsS3Service/TestAmazonS3 stuff is used.

I would wait for @dadoonet to provide details before spending much more
time on mocking things that may not be relevant.

Sorry for the changes; thank you for your effort on this!



On Sun, Jun 12, 2016 at 6:51 PM, gfyoung <notifications@github.com> wrote:

> In
> plugins/repository-s3/src/test/java/org/elasticsearch/cloud/aws/blobstore/S3BlobStoreContainerTests.java
> <https://github.com/elastic/elasticsearch/pull/18815#discussion_r66729195>
> :
>
> > +import java.io.IOException;
> > +import java.util.Locale;
> > +
> > +public class S3BlobStoreContainerTests extends ESBlobStoreContainerTestCase {
> > +    protected BlobStore newBlobStore() throws IOException {
> > +        int maxRetries = 5;
> > +        String region = null;
> > +        Settings settings = Settings.EMPTY;
> > +
> > +        InternalAwsS3Service service = new InternalAwsS3Service(settings);
> > +        AmazonS3 client = service.client(null, Protocol.HTTPS, region, null, null, maxRetries);
> > +        String bucket = randomAsciiOfLength(randomIntBetween(1, 10)).toLowerCase(Locale.ROOT);
> > +
> > +        return new S3BlobStore(settings, client, bucket, region, false,
> > +            new ByteSizeValue(100), maxRetries, "public-read-write", "standard");
> > +    }
>
> @imotov <https://github.com/imotov> : That would be great to have, since
> that would facilitate the test setup for those repositories. However, I
> would still keep mocking on the table as @abeyad
> <https://github.com/abeyad> proposed, since that's what the Google cloud
> tests do with MockHttpTransport.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/elasticsearch/pull/18815/files/5ecfabc0dffffcbec1cbf7c5c7db0e3ab05a8c07#r66729195>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABjkQUhvZEbCRkInDGZvlL7FzEEQ3S6Fks5qLI1vgaJpZM4Iy-tw>
> .
>

tlrx replied:
Concerning Google Cloud Storage: the tests use a `MockHttpTransport` that is dedicated to test Google Client's libraries. It mocks the very last layer of the client, ie HTTP request/response, and avoids to use a real GCS storage account or to set up a testing web server that mocks responses. But the tested code is exactly the same (GCSBlobStore, GCSBlobContainer, Storage client etc) as normal production code.

gfyoung replied:
@abeyad : I did take a look at the `TestAwsS3` and `TestAwsS3Service`, and I actually tried to generate the `BlobStore` based off `TestAwsS3Service`, but I've hit a roadblock here.

abeyad replied:
@gfyoung Whats the road block that you've hit?  Could you share some code?  I think once we find out what the S3 credentials are for testing, it would be a matter of creating an AmazonS3 client, e.g.:
```
AmazonS3 s3Client = new AmazonS3Client(new ProfileCredentialsProvider());        
```

And passing it to the `TestAmazonS3` constructor.  What do you think?

gfyoung replied:
@abeyad : I already have a file up in this PR for these tests.  I just need to know how to properly construct the client.

abeyad replied:
@gfyoung I would suggest simply creating a mock implementation of `AmazonS3` client for now.  This will involve lots of methods (see `AmazonS3Wrapper` which is just a wrapper for a real client, it has all the methods).  The good thing is, you will only have to implement a very select few for now.  I would suggest for the mock client, instead of hitting S3, we should just have an internal `Map` which "stores" the blob name -> blob data.  On this, you can fake out operations such as creating objects, deleting objects, etc. very easily as you're just working with a map.  See https://github.com/elastic/elasticsearch/blob/master/plugins/repository-azure/src/test/java/org/elasticsearch/cloud/azure/storage/AzureStorageServiceMock.java for inspiration.

Does that make sense?

gfyoung replied:
@abeyad : I wish things were this simple, but after a couple of days of attempting to mock the `AmazonS3` client, that just seems a little too difficult.  Mapping the `blobName` to `blobData` would be nice if it weren't for the fact that the `AmazonS3` client also must make mappings from `keyPath + blobName` to `blobData` as well, except that `keyPath` is only accessible to `S3BlobStoreContainer`.  In addition, the `writeBlob` method in `S3BlobStoreContainer` involves no `AmazonS3` client and mocking that is not possible because that is being tested.

abeyad replied:
@gfyoung you could just assume keyPath is always the empty string no matter what is passed in, that would simplify a lot.

> In addition, the writeBlob method in S3BlobStoreContainer involves no AmazonS3 client and mocking that is not possible because that method itself is being tested.

I'm not sure I follow.  `S3BlobStoreContainer#writeBlob` calls `createOutput` which depends on the `S3BlobStore` to create the output stream.  The `AmazonS3` client is part of the `S3BlobStore`.

gfyoung replied:
Yes, I know, but go a little bit deeper, and you'll see that `S3BlobStore` does nothing besides help to instantiate the stream.

abeyad replied:
True, which means you would need a mock implementation of
DefaultS3OutputStream as well - it just needs to be a mock of
java.io.OutputStream that takes your mock AmazonS3 client as a private
member and the write operations of the OutputStream just take the
bytes/input stream and store them in the map of blob data in your mock
AmazonS3.

What do you think?

On Mon, Jun 20, 2016 at 12:10 AM, gfyoung <notifications@github.com> wrote:

> In
> plugins/repository-s3/src/test/java/org/elasticsearch/cloud/aws/blobstore/S3BlobStoreContainerTests.java
> <https://github.com/elastic/elasticsearch/pull/18815#discussion_r67634349>
> :
>
> > +import java.io.IOException;
> > +import java.util.Locale;
> > +
> > +public class S3BlobStoreContainerTests extends ESBlobStoreContainerTestCase {
> > +    protected BlobStore newBlobStore() throws IOException {
> > +        int maxRetries = 5;
> > +        String region = null;
> > +        Settings settings = Settings.EMPTY;
> > +
> > +        InternalAwsS3Service service = new InternalAwsS3Service(settings);
> > +        AmazonS3 client = service.client(null, Protocol.HTTPS, region, null, null, maxRetries);
> > +        String bucket = randomAsciiOfLength(randomIntBetween(1, 10)).toLowerCase(Locale.ROOT);
> > +
> > +        return new S3BlobStore(settings, client, bucket, region, false,
> > +            new ByteSizeValue(100), maxRetries, "public-read-write", "standard");
> > +    }
>
> Yes, I know, but go a little bit deeper, and you'll see that S3BlobStore
> does nothing besides help to instantiate the stream.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/elasticsearch/pull/18815/files/5ecfabc0dffffcbec1cbf7c5c7db0e3ab05a8c07#r67634349>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABjkQQiBXpZ9pD94hIAC0G_7Pkyhvw1-ks5qNhKbgaJpZM4Iy-tw>
> .
>

gfyoung replied:
Well, it's not just `DefaultS3OutputStream` that would I need to mock.  I would to mock `S3BlobContainer` to override `createOutput` as well as `S3BlobStore` to override which `S3BlobContainer` is being created.

gfyoung replied:
Looking again, I don't see how mocking the output stream will help.  The main functionality is wrapped in a `try-block`, with the stream output being on the receiving end of the functionality, not the initiator of it.

gfyoung replied:
I should also that this problem exists with the Azure storage (i.e. `writeBlob` is implemented similarly).

abeyad commented on plugins/repository-azure/src/test/java/org/elasticsearch/repositories/azure/AzureBlobStoreContainerTests.java:
```diff
@@ -0,0 +1,58 @@
+            return blobStore;
+        } catch (URISyntaxException | StorageException e) {
+            throw new IOException(e);
+        }
```
Ditto for my comment on S3 tests..

tlrx commented on plugins/repository-azure/src/test/java/org/elasticsearch/repositories/azure/AzureBlobStoreContainerTests.java:
```diff
@@ -0,0 +1,58 @@
+
+/**
+ * You must specify {@code -Dtests.thirdparty=true -Dtests.config=/path/to/elasticsearch.yml}
+ * in order to run these tests.
```
Is it still necessary?

gfyoung replied:
Depends on whether or not we'll need credentials for this bout of testing AFAIU.  I was modelling off the `AbstractAzureWithThirdPartyIntegTestCase.java` class, so not sure ATM.

tlrx commented on test/framework/src/main/java/org/elasticsearch/repositories/ESBlobStoreContainerTestCase.java:
```diff
@@ -112,6 +112,22 @@ public void testMoveAndList() throws IOException {
         }
     }
 
+    public void testDeleteBlob() throws IOException {
```
Nice!

gfyoung replied:
Thanks! :smile: I should note that this test passes for all of the `BlobContainer` classes that were already being tested as well as Google.

gfyoung commented:
Is there any update on whether or not we can use actual repositories for both Azure or S3 for these tests?  If that can be done, that will save the working  of having to mock everything for both repositories.

jasontedor commented:
> Is there any update on whether or not we can use actual repositories for both Azure or S3 for these tests? 

We can not. These tests need to be able to run from developer machines.

gfyoung commented:
@jasontedor : Ah, I see.  Hmm...then perhaps some more input from others on how to possibly mock the Azure and S3 repositories?  The mocking strategy suggested by @abeyad looked promising, but further discussion of it appears to me that it will only hit a dead end AFAICT.

gfyoung commented:
Sometimes a fresh look at a problem is needed.  Took a different approach to mocking S3-related components and found that I really just needed to mock the client as @abeyad had mentioned at some point.  Onto `Azure` and `HDFS`!

gfyoung commented:
`AzureStorageServiceMock` really made things a lot easier for the `AzureBlobStoreContainer` tests than I had perceived previously.  Only `HDFS` is left now!

gfyoung commented:
Having some issues with `HDFS`: instantiating the `BlobStore` results in a permissions error (even when I create a temp directory that I then instantiate `FileContext` with).

An exception is thrown because it cannot call `UserGroupInformation.getCurrentUser()` via `Subject.getSubject()` (in Hadoop source code) due to permission denials.  How to get around this?

abeyad commented:
@gfyoung great news, let me know when its ready and I can give it another review.

gfyoung commented:
@abeyad : read my comment above - some assistance on the `HDFS` permissions issue would be great!

abeyad commented:
Not familiar with HDFS but perhaps @jbaiera could help?

jbaiera commented:
@abeyad Re HDFS: Many of the Hadoop services' authentication is based around Hadoop's `UserGroupInformation` object, which is poked around with even if auth is turned off. When you first request the `UserGroupInformation` object in the Hadoop Client, it goes through it's internal setup steps. One of the steps is registering with a metrics gathering system so that the UGI can emit quantiles for number of login successes and failures. This is for the benefit of reporting and monitoring systems like Ganglia. The snag occurs when UGI registers it's metrics object. The system tries accessing the `UgiMetrics` object's fields using reflection to collect them as metrics. When the wrapper code asks for all declared fields on `UgiMetrics` it trips on the SecurityManager as it does not have the permission for `accessDeclaredMembers`. I'm afraid I'm not too familiar with the Security side of Java to help much further than that...

gfyoung commented:
Hmmm...seems like a stripped down mock of the FileContext might be needed but not sure yet how to go about that...

gfyoung commented:
You can pass in `Dtests.security.manager=false` to get past the initial error I was getting, but then it breaks because Java has no permissions to call `chmod` when calling `writeBlob`

jasontedor commented:
> You can pass in `-Dtests.security.manager=false` to get past the initial error I was getting

This isn't going to be accepted, these plugins will run under the security manager (so should be tested under it) and the security manager is non-optional in Elasticsearch 5.0.0.

gfyoung commented:
Ah, okay, back to square one it is...is there anyway I can elevate privileges when making these calls or add some other config that will allow me to do so?

abeyad commented:
@gfyoung I'm not very familiar with how HDFS handles these issues that @jbaiera described, but looking at HdfsBlobStore, the code is already there to handle this very situation you describe: https://github.com/elastic/elasticsearch/blob/master/plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsBlobStore.java#L123

Can you verify that code is being invoked when you instantiate the HdfsBlobStore?

gfyoung commented:
@abeyad : if I use the security manager as @jasontedor described, I can't even get there.  The code throws an exception when I attempt instantiate `HdfsBlobStore` as I described <a href="https://github.com/elastic/elasticsearch/pull/18815#issuecomment-227970125">here</a>.

abeyad commented:
@gfyoung at what point in the code is an exception thrown?  can you share a stack trace?

gfyoung commented:
@abeyad : 
[debug.txt](https://github.com/elastic/elasticsearch/files/332808/debug.txt)

abeyad commented on plugins/repository-hdfs/src/test/java/org/elasticsearch/repositories/hdfs/HdfsBlobStoreContainerTests.java:
```diff
@@ -0,0 +1,36 @@
+
+    @Override
+    protected BlobStore newBlobStore() throws IOException {
+        Path tempDir = Files.createTempDirectory(null);
```
use `createTempDir()` instead (its part of `ESTestCase` which `ESBlobStoreContainerTestCase` inherits from)

abeyad commented on plugins/repository-hdfs/src/test/java/org/elasticsearch/repositories/hdfs/HdfsBlobStoreContainerTests.java:
```diff
@@ -0,0 +1,36 @@
+    @Override
+    protected BlobStore newBlobStore() throws IOException {
+        Path tempDir = Files.createTempDirectory(null);
+        return new HdfsBlobStore(FileContext.getFileContext(tempDir.toUri()), "temp", 100);
```
See https://github.com/elastic/elasticsearch/blob/master/plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsRepository.java#L99 for how to create the FileContext with the correct permissions

gfyoung replied:
I did as much boilerplate as I could for the time being, but I think there are some other configurations that I am missing from <a href="https://github.com/elastic/elasticsearch/blob/master/plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsRepository.java#L125">here</a> that are causing the exceptions I am seeing <a href="https://github.com/elastic/elasticsearch/pull/18815#issuecomment-228582832">below</a>.

gfyoung commented:
~~~
:buildSrc:compileJava UP-TO-DATE
:buildSrc:compileGroovy UP-TO-DATE
:buildSrc:writeVersionProperties UP-TO-DATE
:buildSrc:processResources UP-TO-DATE
:buildSrc:classes UP-TO-DATE
:buildSrc:jar UP-TO-DATE
:buildSrc:assemble UP-TO-DATE
:buildSrc:compileTestJava UP-TO-DATE
:buildSrc:compileTestGroovy UP-TO-DATE
:buildSrc:processTestResources UP-TO-DATE
:buildSrc:testClasses UP-TO-DATE
:buildSrc:test UP-TO-DATE
:buildSrc:check UP-TO-DATE
:buildSrc:build UP-TO-DATE
=======================================
Elasticsearch Build Hamster says Hello!
=======================================
  Gradle Version        : 2.13
  OS Info               : Linux 3.19.0-61-generic (amd64)
  JDK Version           : Oracle Corporation 1.8.0_91 [Java HotSpot(TM) 64-Bit Server VM 25.91-b14]
  JAVA_HOME             : /usr/lib/jvm/java-8-oracle
:core:compileJava UP-TO-DATE
:core:generateModulesList UP-TO-DATE
:core:generatePluginsList UP-TO-DATE
:core:processResources UP-TO-DATE
:core:classes UP-TO-DATE
:core:jar UP-TO-DATE
:client:rest:compileJava UP-TO-DATE
:client:rest:processResources UP-TO-DATE
:client:rest:classes UP-TO-DATE
:client:rest:jar UP-TO-DATE
:plugins:repository-hdfs:compileJava UP-TO-DATE
:plugins:repository-hdfs:processResources UP-TO-DATE
:plugins:repository-hdfs:classes UP-TO-DATE
:plugins:repository-hdfs:copyPluginPropertiesTemplate
:plugins:repository-hdfs:pluginProperties UP-TO-DATE
:test:framework:compileJava UP-TO-DATE
:test:framework:processResources UP-TO-DATE
:test:framework:classes UP-TO-DATE
:test:framework:jar UP-TO-DATE
:plugins:repository-hdfs:compileTestJava
:plugins:repository-hdfs:processTestResources UP-TO-DATE
:plugins:repository-hdfs:testClasses
:plugins:repository-hdfs:test
   [junit4] <JUnit4> says hello! Master seed: 7222BA4654ABC020
==> Test Info: seed=7222BA4654ABC020; jvm=1; suite=1
Suite: org.elasticsearch.repositories.hdfs.HdfsBlobStoreContainerTests
  1> [2016-06-25 23:45:08,938][WARN ][org.elasticsearch.bootstrap] Unable to lock JVM Memory: error=12, reason=Cannot allocate memory
  1> [2016-06-25 23:45:08,939][WARN ][org.elasticsearch.bootstrap] This can result in part of the JVM being swapped out.
  1> [2016-06-25 23:45:08,939][WARN ][org.elasticsearch.bootstrap] Increase RLIMIT_MEMLOCK, soft limit: 65536, hard limit: 65536
  1> [2016-06-25 23:45:08,940][WARN ][org.elasticsearch.bootstrap] These can be adjusted by modifying /etc/security/limits.conf, for example: 
  1> 	# allow user 'LINUX-USER-1' mlockall
  1> 	LINUX-USER-1 soft memlock unlimited
  1> 	LINUX-USER-1 hard memlock unlimited
  1> [2016-06-25 23:45:08,940][WARN ][org.elasticsearch.bootstrap] If you are logged in interactively, you will have to re-login for the new limits to take effect.
  2> SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
  2> SLF4J: Defaulting to no-operation (NOP) logger implementation
  2> SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
  2> REPRODUCE WITH: gradle :plugins:repository-hdfs:test -Dtests.seed=7222BA4654ABC020 -Dtests.class=org.elasticsearch.repositories.hdfs.HdfsBlobStoreContainerTests -Dtests.method="testWriteRead" -Dtests.security.manager=true -Dtests.locale=es-GT -Dtests.timezone=Pacific/Efate
ERROR   0.51s | HdfsBlobStoreContainerTests.testWriteRead <<< FAILURES!
   > Throwable #1: java.security.AccessControlException: access denied ("java.io.FilePermission" "/home/LINUX-USER-1/Repositories/elasticsearch/plugins/repository-hdfs/build/testrun/test/J0" "read")
   > 	at __randomizedtesting.SeedInfo.seed([7222BA4654ABC020:83AF2BFFD416B810]:0)
   > 	at java.security.AccessControlContext.checkPermission(AccessControlContext.java:472)
   > 	at java.security.AccessController.checkPermission(AccessController.java:884)
   > 	at java.lang.SecurityManager.checkPermission(SecurityManager.java:549)
   > 	at java.lang.SecurityManager.checkRead(SecurityManager.java:888)
   > 	at java.io.File.exists(File.java:814)
   > 	at org.apache.hadoop.fs.RawLocalFileSystem.mkdirsWithOptionalPermission(RawLocalFileSystem.java:517)
   > 	at org.apache.hadoop.fs.RawLocalFileSystem.mkdirs(RawLocalFileSystem.java:504)
   > 	at org.apache.hadoop.fs.FileSystem.primitiveMkdir(FileSystem.java:1064)
   > 	at org.apache.hadoop.fs.DelegateToFileSystem.mkdir(DelegateToFileSystem.java:161)
   > 	at org.apache.hadoop.fs.FilterFs.mkdir(FilterFs.java:197)
   > 	at org.apache.hadoop.fs.FileContext$4.next(FileContext.java:730)
   > 	at org.apache.hadoop.fs.FileContext$4.next(FileContext.java:726)
   > 	at org.apache.hadoop.fs.FSLinkResolver.resolve(FSLinkResolver.java:90)
   > 	at org.apache.hadoop.fs.FileContext.mkdir(FileContext.java:733)
   > 	at org.elasticsearch.repositories.hdfs.HdfsBlobStore$2.run(HdfsBlobStore.java:66)
   > 	at org.elasticsearch.repositories.hdfs.HdfsBlobStore$2.run(HdfsBlobStore.java:63)
   > 	at org.elasticsearch.repositories.hdfs.HdfsBlobStore$4.run(HdfsBlobStore.java:136)
   > 	at java.security.AccessController.doPrivileged(Native Method)
   > 	at java.security.AccessController.doPrivileged(AccessController.java:713)
   > 	at org.elasticsearch.repositories.hdfs.HdfsBlobStore.execute(HdfsBlobStore.java:133)
   > 	at org.elasticsearch.repositories.hdfs.HdfsBlobStore.mkdirs(HdfsBlobStore.java:63)
   > 	at org.elasticsearch.repositories.hdfs.HdfsBlobStore.<init>(HdfsBlobStore.java:56)
   > 	at org.elasticsearch.repositories.hdfs.HdfsBlobStoreContainerTests.newBlobStore(HdfsBlobStoreContainerTests.java:53)
   > 	at org.elasticsearch.repositories.ESBlobStoreContainerTestCase.testWriteRead(ESBlobStoreContainerTestCase.java:49)
   > 	at java.lang.Thread.run(Thread.java:745)
  2> NOTE: leaving temporary files on disk at: /home/LINUX-USER-1/Repositories/elasticsearch/plugins/repository-hdfs/build/testrun/test/J0/temp/org.elasticsearch.repositories.hdfs.HdfsBlobStoreContainerTests_7222BA4654ABC020-001
  2> NOTE: test params are: codec=Asserting(Lucene60): {}, docValues:{}, maxPointsInLeafNode=1614, maxMBSortInHeap=7.760738480461883, sim=ClassicSimilarity, locale=es-GT, timezone=Pacific/Efate
  2> NOTE: Linux 3.19.0-61-generic amd64/Oracle Corporation 1.8.0_91 (64-bit)/cpus=4,threads=1,free=452914336,total=514850816
  2> NOTE: All tests run in this JVM: [HdfsBlobStoreContainerTests]
Completed [1/1] in 2.27s, 1 test, 1 error <<< FAILURES!

Tests with failures:
  - org.elasticsearch.repositories.hdfs.HdfsBlobStoreContainerTests.testWriteRead

   [junit4] JVM J0:     0.66 ..     3.76 =     3.10s
   [junit4] Execution time total: 3.80 sec.
   [junit4] Tests summary: 1 suite, 1 test, 1 error
:plugins:repository-hdfs:test FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':plugins:repository-hdfs:test'.
> There were test failures: 1 suite, 1 test, 1 error [seed: 7222BA4654ABC020]

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

Total time: 34.647 secs

BUILD SUCCESSFUL
Total time: 4 seconds
~~~
Attached the output as a file for future reference too: 
[debug.txt](https://github.com/elastic/elasticsearch/files/333412/debug.txt)

abeyad commented on plugins/repository-hdfs/src/test/java/org/elasticsearch/repositories/hdfs/HdfsBlobStoreContainerTests.java:
```diff
@@ -18,16 +18,78 @@
+    public FileContext createContext(URI uri) {
+        Configuration cfg = new Configuration(true);
+        cfg.setClassLoader(HdfsRepository.class.getClassLoader());
+        cfg.reloadConfiguration();
```
@costin @jbaiera Do you have any insights as to what should go in the configuration properties for this test to allow it to create an HDFS FileContext for a temporary directory location (created via `createTempDir()`) for this unit test?  This `createContext` method is missing these lines that occur in `HdfsRepository`, but I'm not sure what values should go into them:
```
Map<String, String> map = repositorySettings.settings().getByPrefix("conf.").getAsMap();
        for (Entry<String, String> entry : map.entrySet()) {
            cfg.set(entry.getKey(), entry.getValue());
        }```

costin replied:
Sorry, I don't follow.

The iteration over `conf.` properties is a way for the plugin to enrich the created `Configuration` file (see the plugin documentation for an example).

As for the initial question - what do you mean by "allow it to create"? `Configuration` and `FileContext` are just wrappers.

gfyoung replied:
@costin: What I think @abeyad means is that are there any other configurations that I need to add to avoid that permissions issue?  Also, could you point me to the documentation for the configuration (I presume it is some file in the codebase).

Also, I don't recall seeing a way to initialize a tempdir with the proper permissions.  This current location is within this runtests folder hidden within the codebase locally.

costin replied:
I'm afraid I can't be of much assistance right now. The HDFS integration tests are running against a miniHDFS cluster which is triggered as a fixture (thanks to Ryan and RobertM). Trying to mock Hadoop is likely to be unsuccessful since it's quite greedy in terms of permissions (last I checked, when pointing to a folder, it tried to read the whole tree from the root...).
I think there's still a `TestingFS` that tries to alleviate some of the problems but I'm not sure to what degree it is used.
Double check the permissions created at startup - you might need to create a folder _inside_ the temp folder.

gfyoung replied:
I considered `TestingFS` before, but I was not sure how to incorporate it.  I looked at `HdfsTests` with fresh eyes and found a way.

gfyoung commented:
Finally!  Tests have been successfully added for **ALL** relevant `BlobContainers`.  This is ready for review.

abeyad commented on core/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java:
```diff
@@ -85,7 +85,10 @@ public FsBlobContainer(FsBlobStore blobStore, BlobPath blobPath, Path path) {
-        Files.deleteIfExists(blobPath);
+        if (!Files.deleteIfExists(blobPath)) {
+            // blobPath does not exist
+            throw new IOException("File " + blobPath.toString() + " does not exist");
```
put [ and ] surrounding the blobPath, so `"File [" + blobPath.toString() + "] ...`

gfyoung replied:
Yep, will do.

abeyad commented on plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/blobstore/AzureBlobContainer.java:
```diff
@@ -70,6 +70,11 @@ public boolean blobExists(String blobName) {
+        if (!blobExists(blobName)) {
+            throw new IOException("Blob [" + blobName + "] does not exist");
+        }
+
```
I'm wondering if this is this needed because just below, we catch a `HTTP_NOT_FOUND` and rethrow it as a `FileNotFoundException`

abeyad replied:
I forgot, Azure was ambiguous about the behavior, so its good to keep the blobExists check as you have it

gfyoung replied:
Indeed, tests failed without that check.

abeyad commented on plugins/repository-azure/src/test/java/org/elasticsearch/cloud/azure/storage/AzureStorageServiceMock.java:
```diff
@@ -95,13 +95,13 @@ public OutputStream getOutputStream(String account, LocationMode mode, String co
         for (String blobName : blobs.keySet()) {
             final String checkBlob;
-            if (keyPath != null) {
+            if (keyPath != null || keyPath.isEmpty()) {
```
I don't follow this logic, don't you want `keyPath != null && !keyPath.isEmpty())`?

gfyoung replied:
Yeah, me neither.  Let me go change that.

abeyad commented on plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsBlobContainer.java:
```diff
@@ -68,16 +68,16 @@ public Boolean run(FileContext fileContext) throws IOException {
+            public Boolean run(FileContext fileContext) throws IOException {
+                return fileContext.delete(new Path(path, blobName), true);
+            }
+        });
```
Since HDFS throws the `FileNotFoundException`, we can rely on that and throw it, instead of checking blobExists, no?

gfyoung replied:
Sadly not the case.  That's why I had to add the check in my last commit.

abeyad commented on plugins/repository-hdfs/src/test/java/org/elasticsearch/repositories/hdfs/HdfsBlobStoreContainerTests.java:
```diff
@@ -0,0 +1,103 @@
+        // set file system to TestingFs to avoid a bunch of security
+        // checks, similar to what is done in HdfsTests.java
+        cfg.set(String.format("fs.AbstractFileSystem.%s.impl", uri.getScheme()),
+                TestingFs.class.getName());
```
nice!

gfyoung replied:
I had considered it before, abandoned it, and then @jbaiera 's suggestion caused me to reconsider it and then I found a way, finally! :smile:

abeyad commented on plugins/repository-hdfs/src/test/java/org/elasticsearch/repositories/hdfs/HdfsBlobStoreContainerTests.java:
```diff
@@ -0,0 +1,103 @@
+                    public HdfsBlobStore run() {
+                        try {
+                            FileContext fileContext = createContext(new URI("hdfs:///"));
+                            return new HdfsBlobStore(fileContext, "temp", 100);
```
why 100 bytes as the buffer size?  what about at least 1024?

gfyoung replied:
Really just arbitrary.  I can change to 1024 though.

abeyad commented on plugins/repository-s3/src/test/java/org/elasticsearch/cloud/aws/blobstore/MockAmazonS3.java:
```diff
@@ -0,0 +1,200 @@
+
+class MockAmazonS3 extends AbstractAmazonS3 {
+
+    private Map<String, InputStream> blobs = new ConcurrentHashMap<>();
```
I'm concerned about leaking resources here, is this an issue?  One suggestion is to add a cleanup method that removes the streams from the map and closes them, and then call this cleanup method from an @ AfterClass method in `S3BlobStoreContainerTests`

gfyoung replied:
I don't believe that that is an issue AFAICT, so I would probably leave everything as is.

abeyad commented:
@gfyoung This is looking great!  Thanks for all your hard work and persistence.  I left a few comments.

gfyoung commented:
@abeyad : responded to your round of comments

gfyoung commented:
@abeyad : responded to your round of comments

abeyad commented:
@gfyoung LGTM, thanks for your hard work!

gfyoung commented:
@abeyad : Can this be merged then?

abeyad commented:
@gfyoung Yes

gfyoung commented:
Great!  Can one of the maintainers then merge this if there are no other issues?

rmuir commented on core/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java:
```diff
@@ -85,7 +85,10 @@ public FsBlobContainer(FsBlobStore blobStore, BlobPath blobPath, Path path) {
+        if (!Files.deleteIfExists(blobPath)) {
+            // blobPath does not exist
+            throw new IOException("File [" + blobPath.toString() + "] does not exist");
+        }
```
Instead of making up our own exception, why not use just use Files.delete? This will give you a better exception message. https://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html#delete(java.nio.file.Path)

abeyad replied:
@gfyoung @rmuir's suggestion here is correct.  If you can fix that, I'll merge the PR; sorry I was a bit delayed today in responding.

gfyoung replied:
Fair enough.  Done.

gfyoung commented:
@abeyad : Change has been made.  Ready to merge if there are no other concerns.

javanna commented:
@abeyad I reverted this commit as it was causing consistent test failures, here is an example:

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-intake/942/console

gfyoung commented:
@javanna @abeyad : 

IMHO reverting the entire commit was a bit overkill.  In such situations, I have seen maintainers simply disable tests temporarily with the expectation that the committer (in this case, me) would address them.  Almost never have I seen anyone shy away from the task.

A closer look at the test failures suggests that we have a conflict of behaviours here.

The two test failures are coming from `SharedClusterSnapshotRestoreIT.java` at the following tests:

1) `testDeleteSnapshotWithMissingMetadata`
2) `testDeleteSnapshotWithCorruptedSnapshotFile`

Both test failures was triggered by the change from `Files.deleteIfExists` to `Files.delete` in `FsBlobContainer.java`, which is a lot stricter than `Files.deleteIfExists` because it also throws a `NoSuchFileException`, which is what is thrown in both cases.

In the first test, because the metadata file is deleted, when `FsBlobStoreContainer` tries delete it again (during the snapshot deletion), it can no longer do it.  In the second test, the file corruption makes it unreadable, therefore leading to the exception once again.

Thus, the question is, what needs to change?  Does `FsBlobStoreContainer` need to be more lenient again?  Or do changes need to be made to the snapshot deletion behaviour (where is the code for that BTW)?

javanna commented:
yes @gfyoung I could have probably annotated the test with `@AwaitsFix` and opened an issue about it. I went for baking it off as the PR was very recently merged, those tests were consistently failing and the whole team couldn't run tests hence push anything tested. Don't worry though, once those tests run successfully the PR will for sure be merged back.

abeyad commented:
> IMHO reverting the entire commit was a bit overkill. In such situations, I have seen maintainers simply disable tests temporarily with the expectation that the committer (in this case, me) would address them. Almost never have I seen anyone shy away from the task.

A commit that causes test failures generally means a bug somewhere that was not thought through.  Anyone could create a snapshot build at any time off of master - we don't want to risk master introducing new bugs from a recent PR, to the extent possible, and tests failing is an indication of it.  Hence, I believe @javanna made the right call.  Once its resolved, the everything will be merged back in, so its not an issue.

> testDeleteSnapshotWithMissingMetadata

We should call `expectThrows(IOException.class, () -> client.admin().cluster().prepareDeleteSnapshot("test-repo", "test-snap-1").get())`

The line below it asserting the snapshot is deleted will probably not be true too, you'll have to check and change the assertion in that case as well.

> testDeleteSnapshotWithCorruptedSnapshotFile

Whats the stack trace here?  If corrupted files (which can happen) cause the entire snapshot to be undeletable, then I think we have a problem.  If that's the case, we should open a separate issue indicating snapshot/restore has to handle corrupted files more robustly, and put an `@AwaitsFix` on this test linking to the newly created github issue.  

The thing is, if the `snap-*.dat` file is corrupted, that contains all the info about everything else in the snapshot, so in this test's case, I think it *should* throw an exception if its undeletable due to corruption.  However, if other actual snapshot files could not be deleted due to corruption, but the rest of the snapshot's files could be deleted, I think we still want to get delete the snapshot and all the files we could delete.

abeyad commented:
@gfyoung Also, run `gradle clean check` for everything once your commit is ready, that will run all of the tests (including the ones for the plugins, etc)

gfyoung commented:
Actually, I found the code (`BlobStoreRepository.java`), but still would like some guidance on how to proceed in light of this.

I am less excited about the option now of making `FsBlobContainer` more lenient because `BlobStoreIndexShardRepository` does take a generic `BlobContainer`, meaning that making this blob container more lenient would require making every `BlobContainer` implemented less lenient, which actually is pretty much reverting my entire PR. (so may actually not be overkill)

The other suggestion is to follow what `BlobStoreIndexShardRepository.java` does by catching the `IOException` and then logging it with `logger.debug`.  This option will also "resolve" those failures.

gfyoung commented:
What about following what `BlobStoreIndexShardRepository.java` does by catching the `IOException` on each deletion call and then logging it with `logger.debug`?

gfyoung commented:
Also, your proposed change for `testDeleteSnapshotWithMissingMetadata` does not appear to work, at least when I run tests locally.

abeyad commented:
@gfyoung i'll pull down your branch and experiment

abeyad commented:
@gfyoung I've created a branch here: https://github.com/abeyad/elasticsearch/tree/stricter-delete-blob.  I'd like to issue a PR against your branch, but I don't have access to your elasticsearch fork.

Before incorporating, I'd like feedback from @imotov first on what I've done.

gfyoung commented:
@abeyad : My original branch is <a href="https://github.com/gfyoung/elasticsearch/tree/stricter-delete-blob">here</a>.  You should be able to access that hopefully.  Thanks for doing this!

abeyad commented:
@gfyoung I haven't forgotten about this; I've discussed some changes with @imotov that should be done in order to avoid issues revealed with the two broken tests.  I'm working on those now and will hopefully have a PR for review in the next couple days.  Then both our work can be merged together.

Can I request that you remove your latest commit to this branch and push?  My commits will cause merge conflicts with the latest commit.  

Lastly, I still can't issue a PR against your branch.  I believe you need to add me as one of your collaborators for elasticsearch (https://github.com/gfyoung/elasticsearch/settings/collaboration).  Alternatively, we can setup a feature branch that incorporates both of our commits - this may be the best approach so I'll set it up and let you know when its there so you can push your commits to it.

gfyoung commented:
@abeyad : I reverted that last commit on that branch.  That was my attempt to patch the errors.  Strange that you can't put up PR's against my fork (I haven't seen that issue before), but yes, do let me know once the feature branch is up!

gfyoung commented:
@abeyad : any updates on the feature branch?
